src/main/java/net/fabricmc/loader/metadata/V0ModMetadataParser.java

```diff
@@ -0,0 +1,430 @@
+
+package net.fabricmc.loader.metadata;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import net.fabricmc.loader.api.Version;
+import net.fabricmc.loader.api.VersionParsingException;
+import net.fabricmc.loader.api.metadata.ContactInformation;
+import net.fabricmc.loader.api.metadata.ModDependency;
+import net.fabricmc.loader.api.metadata.ModEnvironment;
+import net.fabricmc.loader.api.metadata.Person;
+import net.fabricmc.loader.lib.gson.JsonReader;
+import net.fabricmc.loader.lib.gson.JsonToken;
+import net.fabricmc.loader.util.version.VersionDeserializer;
+
+final class V0ModMetadataParser {
+	private static final Pattern WEBSITE_PATTERN = Pattern.compile("\\((.+)\\)");
+	private static final Pattern EMAIL_PATTERN = Pattern.compile("<(.+)>");
+
+	public static LoaderModMetadata parse(JsonReader reader) throws IOException, ParseMetadataException {
+		// All the values the `fabric.mod.json` may contain:
+		// Required
+		String id = null;
+		Version version = null;
+
+		// Optional (mod loading)
+		Map<String, ModDependency> requires = new HashMap<>();
+		Map<String, ModDependency> conflicts = new HashMap<>();
+		V0ModMetadata.Mixins mixins = null;
+		ModEnvironment environment = ModEnvironment.UNIVERSAL; // Default is always universal
+		String initializer = null;
+		List<String> initializers = new ArrayList<>();
+
+		String name = null;
+		String description = null;
+		Map<String, ModDependency> recommends = new HashMap<>();
+		List<Person> authors = new ArrayList<>();
+		List<Person> contributors = new ArrayList<>();
+		ContactInformation links = null;
+		String license = null;
+
+		while (reader.hasNext()) {
+			switch (reader.nextName()) {
+			case "schemaVersion":
+				// Duplicate field, make sure it matches our current schema version
+				if (reader.peek() != JsonToken.NUMBER) {
+					throw new ParseMetadataException("Duplicate \"schemaVersion\" field is not a number", reader);
+				}
+
+				final int read = reader.nextInt();
+
+				if (read != 0) {
+					throw new ParseMetadataException(String.format("Duplicate \"schemaVersion\" field does not match the predicted schema version of 0. Duplicate field value is %s", read), reader);
+				}
+
+				break;
+			case "id":
+				if (reader.peek() != JsonToken.STRING) {
+					throw new ParseMetadataException("Mod id must be a non-empty string with a length of 3-64 characters.", reader);
+				}
+
+				id = reader.nextString();
+				break;
+			case "version":
+				if (reader.peek() != JsonToken.STRING) {
+					throw new ParseMetadataException("Version must be a non-empty string", reader);
+				}
+
+				try {
+					version = VersionDeserializer.deserialize(reader.nextString());
+				} catch (VersionParsingException e) {
+					throw new ParseMetadataException("Failed to parse version", e);
```

Comment:
this should store the raw version in a variable, then add the wrong one to the exc